src/netaddress.cpp

```diff
@@ -101,24 +193,76 @@ bool CNetAddr::SetInternal(const std::string &name)
     return true;
 }
 
+namespace torv3 {
+// https://gitweb.torproject.org/torspec.git/tree/rend-spec-v3.txt#n2135
+static constexpr size_t CHECKSUM_LEN = 2;
+static constexpr size_t VERSION_LEN = 1;
+static const unsigned char VERSION[] = "\x03";
+static constexpr size_t TOTAL_LEN = ADDR_TORV3_SIZE + CHECKSUM_LEN + VERSION_LEN;
+
+static void Checksum(Span<const uint8_t> addr_pubkey, uint8_t (&checksum)[CHECKSUM_LEN])
+{
+    // CHECKSUM = H(".onion checksum" | PUBKEY | VERSION)[:2]
+    static const unsigned char prefix[] = ".onion checksum";
+    static constexpr size_t prefix_len = 15;
+
+    SHA3_256 hasher;
+
+    hasher.Write(MakeSpan(prefix).first(prefix_len));
+    hasher.Write(addr_pubkey);
+    hasher.Write(MakeSpan(torv3::VERSION).first(torv3::VERSION_LEN));
+
+    uint8_t checksum_full[SHA3_256::OUTPUT_SIZE];
+
+    hasher.Finalize(checksum_full);
+
+    memcpy(checksum, checksum_full, sizeof(checksum));
+}
+
+}; // namespace torv3
+
 /**
  * Parse a TORv2 address and set this object to it.
  *
  * @returns Whether or not the operation was successful.
  *
  * @see CNetAddr::IsTor()
  */
-bool CNetAddr::SetSpecial(const std::string &strName)
+bool CNetAddr::SetSpecial(const std::string &str)
 {
-    if (strName.size()>6 && strName.substr(strName.size() - 6, 6) == ".onion") {
-        std::vector<unsigned char> vchAddr = DecodeBase32(strName.substr(0, strName.size() - 6).c_str());
-        if (vchAddr.size() != ADDR_TORV2_SIZE) {
+    static const char* suffix = ".onion";
+    static constexpr size_t suffix_len = 6;
+
+    if (str.size() <= suffix_len || str.substr(str.size() - suffix_len) != suffix) {
+        return false;
+    }
+
+    const auto& input = DecodeBase32(str.substr(0, str.size() - suffix_len).c_str());
```

Comment:
Hmm, both `c_str()` and `data()` return the same thing. I kept it to `c_str()` because `DecodeBase32()` expects a nul-terminated string and `c_str()` has always been returning such string, whereas `data()` does so only after C++11. I.e. `c_str()` is more widely known to return a nul-terminated string.

Anyway, I added a check whether the string contains an embedded `'\0'` plus a test for it.